fdbserver/workloads/KVStoreTest.actor.cpp

```diff
@@ -252,6 +303,204 @@ struct KVStoreTestWorkload : TestWorkload {
+	}
+	TraceEvent("KVStoreRangeReadClearCount").detail("Count", count);
+
+	rowLimit = -1000;
+	key = LiteralStringRef("\xff\xff\xff\xff");
+	count = 0;
+
+	// same logic as above except we reverse the direction of the range read results (desc order)
+	while (true) {
+		Key start;
+		state KeyRangeRef range2 = KeyRangeRef(start, key);
+		state Standalone<RangeResultRef> kv2 = wait(test.store->readRange(range2, rowLimit, byteLimt));
+		state Reference<ClearRangeResultWriter> resultWriter2 =
+		    makeReference<ClearRangeResultWriter>(rowLimit, byteLimt);
+
+		if (kv2.size() == 0)
+			break;
+		int rand1 = deterministicRandom()->randomInt(0, kv2.size()),
+		    rand2 = deterministicRandom()->randomInt(0, kv2.size());
+		state int clearStartIdx2 = std::min(rand1, rand2);
+		state int clearEndIdx2 = std::max(rand1, rand2);
+		resultWriter2->setClearRange(kv2[clearEndIdx2].key, kv2[clearStartIdx2].key);
+		wait(
+		    test.store->readRange(KeyRangeRef(kv2[kv2.size() - 1].key, keyAfter(kv2[0].key)), resultWriter2, rowLimit));
+		int curIdx = 0;
+		ASSERT(resultWriter2->result.size() == kv2.size() - (clearEndIdx2 - clearStartIdx2));
+		for (int i = 0; i < kv2.size(); i++) {
+			if (i <= clearStartIdx2 || i > clearEndIdx2) {
+				ASSERT(kv2[i].key == resultWriter2->result[curIdx].key);
+				ASSERT(kv2[i].value == resultWriter2->result[curIdx].value);
+				curIdx++;
+			}
+		}
+		ASSERT(curIdx == resultWriter2->result.size());
+		count += resultWriter2->result.size();
+		if (kv2.size() < -rowLimit)
+			break;
+		key = kv2[kv2.size() - 1].key;
+	}
+	TraceEvent("KVStoreRangeReadClearCountRev").detail("Count", count);
+
+	return Void();
+}
+
+ACTOR Future<Void> testReadRangeEdgeCases(KVStoreTestWorkload* workload, KVTest* ptest) {
+	/*
+	  - Tests various edge cases with the new result writer range read interface
+	*/
+	state KVTest& test = *ptest;
+	state Key key;
+	state int limit = 1000;
+	state int byteLimt = 1 << 30;
+	state int curIdx = 0;
+	state KeyRangeRef range = KeyRangeRef(key, LiteralStringRef("\xff\xff\xff\xff"));
+	state Standalone<RangeResultRef> kv = wait(test.store->readRange(range, limit, byteLimt));
+	state Reference<ClearRangeResultWriter> resultWriter = makeReference<ClearRangeResultWriter>(limit, byteLimt);
+
+	// Tests with a clear range at the begining of the read range
+	resultWriter->setClearRange(kv[0].key, kv[kv.size() / 2 - 1].key);
+	wait(test.store->readRange(KeyRangeRef(kv[0].key, keyAfter(kv[kv.size() - 1].key)), resultWriter, limit));
+	for (int i = 0; i < kv.size(); i++) {
+		if (i >= kv.size() / 2 - 1) {
+			ASSERT(kv[i].key == resultWriter->result[curIdx].key);
+			ASSERT(kv[i].value == resultWriter->result[curIdx].value);
+			curIdx++;
+		}
+	}
+	ASSERT(resultWriter->result.size() > 0);
+	ASSERT(curIdx == resultWriter->result.size());
+
+	resultWriter->clearResults();
+
+	// Tests with a clear range at the end of the read range
+	resultWriter->setClearRange(kv[kv.size() / 2 - 1].key, keyAfter(kv[kv.size() - 1].key));
```

Comment:
Another edge cast to test is what if the entire range you are reading has been cleared.